### PR TITLE
[TopOpt] First draft for reactivation

### PR DESCRIPTION
**Pull request by PhiHo-eng**
**Topology Optimization**
The topology optimization application in Kratos has been reactivated, linked with the structural mechanics application and extended. [I carried this work under the supervision of @e-dub at unibz]

**Changes in the new Topology Optimization Application**
The following has been done:
- Change solving application from SolidMechanicsApplication to StructuralMechanicsApplication
- Update with Pybind11
- Density filtering
- Heaviside projection in the density filter
- Update of the filtering function (now undependent of the Young's modulus)
- MMA and GCMMA as optimization algorithms (original code in external libraries)
- Reactivation of the smoothing function and the STL file as product of the post processing

### CODE REVIEW

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
I've quickly browsed the PR and there are some changes in the `scripts` folder that shouldn't be there. Besides. There is a `run_TopOpt.py` script in the repo main folder that must be there neither.

File: applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 1]
```diff
@@ -30,7 +30,6 @@
 #include "custom_utilities/mapping/mapper_vertex_morphing_improved_integration.h"
 #include "custom_utilities/mapping/mapper_vertex_morphing_symmetric.h"
 #include "custom_utilities/damping/damping_utilities.h"
-#include "custom_utilities/damping/direction_damping_utilities.h"
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp]: There should be no changes made to ShapeOptimizationApplication, i guess something went wrong when you merged.

File: applications/TopologyOptimizationApplication/custom_elements/comparison_utilities.h [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+					 (StressTensor(1,1)-StressTensor(2,2))*((StressTensor(1,1)-StressTensor(2,2)))+
+					 (StressTensor(2,2)-StressTensor(0,0))*((StressTensor(2,2)-StressTensor(0,0)))+
+					 6*(StressTensor(0,1)*StressTensor(1,0)+StressTensor(1,2)*StressTensor(2,1)+StressTensor(2,0)*StressTensor(0,2)));
+
+	if( SigmaEquivalent < 0 )
+	  SigmaEquivalent = 0;
+
+	SigmaEquivalent = sqrt(SigmaEquivalent);
+
+	//in principal stresses:
```
Comment (armingeiser) [inline applications/TopologyOptimizationApplication/custom_elements/comparison_utilities.h:138]: please clean such commented code parts

File: applications/TopologyOptimizationApplication/examples/00_Small_Cantilever_Hexahedral_Mini/README.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		solver_module anstatt import_module
+
+2. 	ProjectParameter:
+		StaticSIMP als solvertype auf Static umgeändert
+
+Versuche es mit der Vorlage aus GitHub das run_TopOpt zusammenzustellen. Bei Versuchen schon weiter gekommen als mit momentaner Variante!
+
+Aufpassen mit Solver. Muss dieser selbst erstellt werden? Oder wird ein bereits vorhanderner Solver verwendet? Ist es das was im Programm steht und warum es nicht weiter geht? Den neuen Solver in python_scripts nochmal anschauen und vergleichen mit den Solver für SructuralMechanicsApplication
+
+
```
Comment (armingeiser) [inline applications/TopologyOptimizationApplication/examples/00_Small_Cantilever_Hexahedral_Mini/README.md]: This example (the mdpa) is too large for the main repo, there is a dedicated Examples repository where such large "use-case" examples should be created.

File: applications/TopologyOptimizationApplication/examples/00_Small_Cantilever_Hexahedral_Mini/run_TopOpt.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
+#import kratos core and applications
+import KratosMultiphysics as km
+import KratosMultiphysics.StructuralMechanicsApplication as ksm
+import KratosMultiphysics.TopologyOptimizationApplication as kto
+import KratosMultiphysics.LinearSolversApplication as kls
+import os
+import OptimizationParameters as opt_parameters
```
Comment (armingeiser) [inline applications/TopologyOptimizationApplication/examples/00_Small_Cantilever_Hexahedral_Mini/run_TopOpt.py]: i suggest to store the opt settings also as a json file, like it is done in the rest of Kratos

File: applications/TopologyOptimizationApplication/examples/00_Small_Cantilever_Hexahedral_Mini/run_TopOpt_PostProcessing.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                                            }''')
+                            )
+gid_io_3.ExecuteInitialize()
+# gid_io_2.ExecuteBeforeSolutionLoop()
+gid_io_3.ExecuteInitializeSolutionStep()
+gid_io_3.PrintOutput()
+gid_io_3.ExecuteFinalizeSolutionStep()
+gid_io_3.ExecuteFinalize()
+
+IOUtilities().WriteSurfaceAsSTLFile("smoothed_design_00_Cantilever.stl",extracted_surface_model_part)
```
Comment (armingeiser) [inline applications/TopologyOptimizationApplication/examples/00_Small_Cantilever_Hexahedral_Mini/run_TopOpt_PostProcessing.py]: the resulting stl file seems to be commited, should definitely not be there.

**Review by armingeiser — CHANGES_REQUESTED** [review]
Thanks for reviving the topology optimization app. Before reviewing in detail i have a few general comments:

- please remove all the examples from this PR:
   - there is an "Examples" repo made for storing those
   - instead create one small test case (e.g. MBB cantilever) with <<1000 nodes. 
   - ideally you set it up as a test case that compares the resulting topology (e.g. comparing the density distribution after the optimization with some reference values. 
- as Ruben has said, there are some files changed/added in folders outside of the application, please revert those changes. 
- Minor cosmetics: ev. you can add a small gif of a topology optimization and a convergence graph to the PR description

Once these things are done, the number of changed files (currently 97) should go down significantly, making the review easier.

**Review by philbucher — CHANGES_REQUESTED** [review]
**Since this is a very large PR I strongly recommend the following:**
Split up the PR into smaller ones, this way you will get a better review and also it will get merged much faster!

First I would create an empty application with the [ApplicationGenerator](https://github.com/KratosMultiphysics/Kratos/tree/master/kratos/python_scripts/application_generator) (=> `python3 createApplication.py TopologyOptimizationApplication`) and add it to the repo in an initial PR 
(then you will also get the latest updates which were done to e.g. `CMakeLists.txt`)
=> this way it can get merged super fast and you will already profit that if sth gets changed in master, your app will also receive the changes

Then afterwards I recommend adding your implementation in smaller chunks so that we can review it easier. Ofc this is a little more effort for you at first, but in the long run this will be faster AND you will get a more detailed review from us with suggestions on how to improve the code

File: applications/TopologyOptimizationApplication/custom_elements/solid_mechanics_math_utilities.h [hunk 6]
```diff
@@ -0,0 +1,1773 @@
+//
+//   Project Name:        KratosSolidMechanicsApplication $
+//   Created by:          $Author:            JMCarbonell $
+//   Last modified by:    $Co-Author:                     $
+//   Date:                $Date:                July 2013 $
+//   Revision:            $Revision:                  0.0 $
+//
+//
+
+#if !defined(KRATOS_SOLID_MECHANICS_MATH_UTILITIES_H_INCLUDED)
```
Comment (armingeiser) [inline applications/TopologyOptimizationApplication/custom_elements/solid_mechanics_math_utilities.h:10]: Do you still need this? Please remove all functions /files that are not needed.

File: applications/TopologyOptimizationApplication/tests/test.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+
```
Comment (armingeiser) [inline applications/TopologyOptimizationApplication/tests/test.txt:1]: leftover

**Pull request closed without merging**